### ISSUE TITLE
UDiT-L is bigger than DiT-XL

### ISSUE THREAD

**Issue by LinB203** [YuchuanTian/U-DiT#5]
Thank you for your great work, I have some suggestions and questions.

In fact, the UDiT-L has more parameters than the DiT-XL, 0.9B vs. 0.6B. It seems unfair to make this comparison. Also maybe some actual training hours could be provided to prove that it is efficient.
Also why did FFN's ratio drop from 4 to 2?

**Comment by YuchuanTian** [comment]
Thank you for your comments. Here are several points I want to make:

1. We align FLOPs for the paper, and it's hard to align FLOPs and params at the same time. Plus, we hold that U-DiT-B (a smaller version) is sufficient to show the advantage of U-DiTs over DiTs with much fewer FLOPs and params.

2. Here we provide training speed info. U-DiT (vanilla) does have a better speed-FID tradeoff compared to DiT-XL/2. Comparing vanilla version of U-DiT against DiT shows the direct advantage of the proposed U-Net architecture.

|                       | Training Speed (steps/sec) | FID   |
| --------------------- | ----------------- | ----- |
| DiT-XL/2*              | 1.71              | 20.05 |
| U-DiT-L (Vanilla)     | 1.55              | 12.04 |
| U-DiT-L (with tricks) | 0.84              | 10.08 |
| U-DiT-B (Vanilla)     | 3.14              | 20.89 |

p.s. For vanilla models, please refer to Table 9 in our [latest manuscript](https://arxiv.org/pdf/2405.02730v2) (updated on 6/3).

3. For mlp ratio, we simply followed previous works on low-level models (in fact, we haven't looked into this issue so far; thanks for raising it). But in toy experiments, we keep mlp ratio=4 for fair comparison.

**Comment by LinB203** [comment]
Also I have another question that I would like your guidance on. When the ratio of downsampler attention=2, the maximum kernel size=5 in the FFN is just enough to cover itself as well as the neighboring 2×2 grids. May I ask if this k=5 is designed this way? If I want to scale up to a higher resolution, say 512, then maybe the ratio of downsampler attention is 4 at this point, which means the maximum kernel size of the FFN should be 9?

**Comment by LinB203** [comment]
For mlp ratio, I found that when keeping comparable flops, mlp ratio=2 is necessary. when keeping mlp ratio=4, the flops are much larger than DiT (also mlp ratio=4), and this is because there are too many convolutions with k>1 in FFN.

I'm worried that it's just an inductive bias due to the convolution. Thinking about it differently, downsampler attention can be viewed as sparse global attention, however it lacks local interactions (a convolution layer in the downsampler is weak), so the FFN acts as a local feature interaction.
Therefore fast convergence is obvious, but is the UDiT framework still superior when the amount of data or model size or number of training steps is large enough. This is the concern.

**Comment by YuchuanTian** [comment]
Thanks for sharing your Pixart results! I am glad our downsampler scheme works on Pixart.

As for kernel size, actually we did not think that further. We simply follow rep tricks that include kersize=1/3/5 branches, and that's why we use kersize=5.

However, I don't think enlarging kersize to 9 is a good idea, as kersize>5 could significantly slow down the model. Another noteworthy thing is that removing this dwconv could cause FID drops (U-DiT with tricks vs. vanilla U-DiT), but the drop is not significant for model as large as U-DiT-L (this probably means FFN dwconv is not *that* important). Hence, I think kersize=5 keeps a good balance between performance and efficiency (and affordable rep tricks are practical as well!)

As for the mlp-ratio issue, the model may have to be completely redesigned when mlp-ratio changes: I think it's necessary to add other model hyperparameters like depth, channels, head num, etc. into design space to align U-DiTs with DiTs in terms of FLOPs. (In fact, we had a hard time trying model hyperparameters for the released configs...)

From the existing experiments, I tend to hold that the convolution inductive bias is indeed helpful, but it is not the determining factor. Removing convolution in FFN does not bring significant performance decay: the advantage of U-DiT is still obvious. 

We conducted experiments lengthening training iterations to 1M and enlarging image size to 512, and both show that U-DiT could outcompete DiT-XL/2 by large margins. Hence, I tend to hold that U-DiT is still superior for large-scale data/models.

**Comment by lh-zhu** [comment]
> Thanks for sharing your Pixart results! I am glad our downsampler scheme works on Pixart.
> 
> As for kernel size, actually we did not think that further. We simply follow rep tricks that include kersize=1/3/5 branches, and that's why we use kersize=5.
> 
> However, I don't think enlarging kersize to 9 is a good idea, as kersize>5 could significantly slow down the model. Another noteworthy thing is that removing this dwconv could cause FID drops (U-DiT with tricks vs. vanilla U-DiT), but the drop is not significant for model as large as U-DiT-L (this probably means FFN dwconv is not _that_ important). Hence, I think kersize=5 keeps a good balance between performance and efficiency (and affordable rep tricks are practical as well!)
> 
> As for the mlp-ratio issue, the model may have to be completely redesigned when mlp-ratio changes: I think it's necessary to add other model hyperparameters like depth, channels, head num, etc. into design space to align U-DiTs with DiTs in terms of FLOPs. (In fact, we had a hard time trying model hyperparameters for the released configs...)
> 
> From the existing experiments, I tend to hold that the convolution inductive bias is indeed helpful, but it is not the determining factor. Removing convolution in FFN does not bring significant performance decay: the advantage of U-DiT is still obvious.
> 
> We conducted experiments lengthening training iterations to 1M and enlarging image size to 512, and both show that U-DiT could outcompete DiT-XL/2 by large margins. Hence, I tend to hold that U-DiT is still superior for large-scale data/models.

Hi, Yuchuan. I am really interested in U-DiT. Could you please share the FID comparison curve between U-DiT-L and DiT-XL from 100K iters to 1M iters? The detailed comparison curves can help us better acknowledge the strengths of U-DiT :)

**Comment by YuchuanTian** [comment]
Well noted. I will try to release them next week. Thank you for your interst!
